### DISCUSSION TITLE
How to assign record ID inside OnRecordCreateRequest hook

### DISCUSSION THREAD

**Discussion by thangngoc89** [pocketbase/pocketbase#6170] [Q&A] [answered]
I'm trying to upgrade my app to 0.23.12 and so far so good. I love this new simplify design, everything just make senses.

However, I'm facing this new challenge where I need the record ID for further processing down the line inside OnRecordCreateRequest hook.

I created a very minimal reproduce app below:

`posts` collection is just an empty collection with id field

```go
package main

import (
	"log"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/security"
)

func main() {
	app := pocketbase.New()

	app.OnRecordCreateRequest("posts").BindFunc(func(e *core.RecordRequestEvent) error {
		if e.Record.Id == "" {
			e.Record.Set("id", security.RandomString(15))
		}

		// further processing that requires e.Record.Id

		return e.Next()
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

Now when I try to create a new posts record with empty `id`, this would be the error:

<img width="699" alt="image" src="https://github.com/user-attachments/assets/fa8b113a-f27b-4877-b761-894e1e605de7" />

```
ERROR POST /api/collections/pbc_1125843985/records
└─ Failed to create record.
   └─ map[data:id: Invalid value format. raw:id: Invalid value format.]
```

Any suggestions on how to do this? Thank you.

**Comment by ganigeorgiev** [comment, marked as answer]
`security.RandomString(15)` generates a random string with characters from a-zA-Z0-9.

The default collection `id` pattern for new record is `^[a-z0-9]+$`, aka. it may fail because of the uppercase letters. You need to either change the pattern from the field option (_I don't recommend allowing uppercase letters because on some systems like Windows may cause issues since the record id is also used as directory for the uploaded files; we even have an extra check to prevent records with duplicated ids in case-insensitive manner for this exact case_) OR a better option is to use the `text` field `:autogenerate` set modifier and leave the field to generate a value for itself based on its "Autogenerate pattern" option, aka. you can try something like:

```
record.Set("id:autogenerate", "")
```
_(see https://pocketbase.io/docs/collections/#textfield)_

**Reply by chnfyi** [reply]
Shouldn't the demo and its image on the landing page be updated, considering that the demo's default records `id` pattern is `[A-Za-z0-9]+` while the default collections `id` pattern is `^[a-z0-9]+$`? They could be misleading regarding your above recommandation.

**Reply by ganigeorgiev** [reply]
> considering that the demo's default records id pattern is [A-Za-z0-9]+ while the default collections id pattern is ^[a-z0-9]+$

You are mistaken, or at least I don't understand what do you mean. The `id` field pattern is `^[a-z0-9]+$`.

Old existing records should still continue to work as it is. The above pattern is just the new default with v0.23 for new records.

_edit: If you mean that the already inserted records are using ids with mixed case then that's fine. We can adjust it but I don't see the point of changing it. In any case my recommendation doesn't matter, the field has a validator option and user can change it if they want to._

**Reply by thangngoc89** [reply]
@ganigeorgiev how would I immediately get the `id` after Set `id:autogenerate` ? I need to access the Id inside the hook before the save operation. Or should I generate the id separately?

```go
record := e.Record
record.Set("id:autogenerated", "")

if record.Id == "" {
    return apis.NewInternalServerError("sequence_runs id is none", record.Id)
}
```

This is what I came up with according to your explanation:

```go
newId := security.RandomStringWithAlphabet(core.DefaultIdLength, core.DefaultIdAlphabet)
record.Set("id", newId)

// accessing newId or record.Id below 
```

I hope that this doesn't come with any negative effects?

**Reply by ganigeorgiev** [reply]
`record.Set("id:autogenerate", "")` should result in the record having an id, but in your case it seems that you have a typo in the modifier name (_remove the last `d`_)

**Reply by thangngoc89** [reply]
@ganigeorgiev thank you so much. That did it!
